commercial/app/model/merchandise/books/BookFinder.scala

```diff
@@ -2,67 +2,56 @@ package commercial.model.merchandise.books
-import common.Logging
+import commercial.model.merchandise.Book
+import common.{AkkaAgent, Logging}
 import conf.Configuration
 import conf.switches.Switches.BookLookupSwitch
-import commercial.model.merchandise.Book
 import play.api.libs.json._
 import play.api.libs.oauth.{ConsumerKey, OAuthCalculator, RequestToken}
 import play.api.libs.ws.{WSClient, WSSignatureCalculator}
-import shade.memcached.{Memcached, MemcachedCodecs, Configuration => MemcachedConfiguration}
 
 import scala.concurrent.duration._
 import scala.concurrent.{ExecutionContext, Future}
-import scala.util.control.NonFatal
 
 class BookFinder(actorSystem: ActorSystem, magentoService: MagentoService) extends Logging {
 
-  private implicit lazy val executionContext = actorSystem.dispatchers.lookup("akka.actor.memcached")
-
-  private lazy val defaultBookDataCache = new MemcachedBookDataCache(actorSystem)
-
-  def findByIsbn(isbn: String,
-                 cache: BookDataCache = defaultBookDataCache,
-                 lookup: String => Future[Option[JsValue]] = magentoService.findByIsbn):
-  Future[Option[Book]] = {
-
-    def cachedBook(bookData: JsValue): Future[Option[Book]] = {
-      Future.successful {
-        bookData match {
-          case JsNull => None
-          case _ =>
-            bookData.validate[Book] fold(
-              invalid => {
-                log.error(Json.stringify(JsError.toJson(invalid)))
-                None
-              },
-              book => Some(book))
-        }
+  private implicit val bookActorExecutionContext: ExecutionContext = actorSystem.dispatchers.lookup("akka.actor.book-lookup")
+  private implicit val bookActorTimeout: Timeout = 0.2.seconds
+  private implicit val magentoServiceImplicit = magentoService
+
+  def findByIsbn(isbn: String): Option[Book] = BookAgent.get(isbn) map { _.as[Book] }
+}
+
+object BookAgent extends Logging {
+
+  private lazy val cache = AkkaAgent(Map.empty[String, JsValue])
+  private lazy val lookupJobs = AkkaAgent(Set.empty[String])
+
+  def get(isbn: String)(implicit magentoService: MagentoService, executionContext: ExecutionContext): Option[JsValue] =
+
+    cache.get.get(isbn) match {
+      case Some(json) => {
+        log.info(s"Cache hit for ISBN: $isbn.")
+        Some(json)
       }
-    }
+      case None => {
+        log.info(s"Cache miss for ISBN: $isbn.")
 
-    lazy val lookedUpBook = {
-      lookup(isbn) map {
-        _ map { bookData =>
-          cache.add(isbn, bookData)
-          bookData.as[Book]
-        } orElse {
-          cache.add(isbn, JsNull)
-          None
+        if (lookupJobs.get.contains(isbn)) {
+          log.info(s"ISBN lookup is already due for $isbn. Not running Magento Service.")
         }
+        else {
+          log.info(s"Looking up ISBN $isbn with Magento Service.")
+          magentoService.findByIsbn(isbn) map (_.foreach { json: JsValue => cache alter { _ + (isbn -> json) } })
+          lookupJobs alter { _ + isbn }
+        }
+        None
```

Comment:
It helped a lot when trying to see where a component was coming from, however it might add a lot of noise to the PROD logs.

What level of logs does Kibana consume? I could make these `log.debug` statements instead or just remove them. If we have any trouble with this functionality in future then presumably whoever looks at it can add their own temporary logging.